ng client improve TrustedMetadataSet testing

MVrachev opened the pull request:
Fixes #1465

**Description of the changes being introduced by the pull request**:
The current situation with the TrustedMetadataSet testing is that
we don't have a minimal amount of unit tests testing the different
branches in the various API functionality in the class.

This commit proposes simple unit tests covering almost all of the
branches in the API functions and increasing the unit test coverage
(as reported from the "coverage" tool) from 74 % to 97 %.

The code could be complicated at places because the different
branches in the update_* functions depend on other metadata classes
as well.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

**History of modifications made on the pr**:

1. An initial version of the pr: 
  1.1 Move the shared code between tests into the "setupClass" function.
  1.2 Add longer test functions with multiple test cases inside them
2. Split longer test functions into multiple smaller ones
  2.1 Added a couple of `setup` helper functions calling `trusted_metadata_set.update_*` 
  aiming to move the common code between tests into separate functions.
3. Addressing multiple suggestions and optimizations by @jku.
4. Experimenting with Jussi suggestion that instead of modifying internal trusted_metadata_set state
we should modify the input data we initially feed for trusted_metadata_set.
  4.1 Add `modify_snapshot`, `modify_timestamp` helper functions splitting common code that
  now can be reused. They end up with quite complicated functions because I wanted to reuse them
  in different scenarios.
5. By suggestion of @jku, replacing `modify_snapshot` and `modify_timestamp` with a 
higher order function `modify_metadata` which accepts a modification function as a parameter 
then calls it. This allows the actual modification function to be defined inside the test itself and 
be a lot simpler and still reuse a lot of the code responsible to serialization, signing and other 
operations inside the `modify_metadata` function.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
ng client metadata set: organize common test code

Move the shared code between tests into the "setupClass" function.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
ngclient TrustedMetadataSet: improve unit testing

The current situation with the TrustedMetadataSet testing is that
we don't have a mnimimal amount of unit tests testing the different
branches in the various API functionality in the class.

This commit proposes simple unit tests covering almost all of the
branches in the API functions and increasing the unit test coverage
(as reported from the "coverage" tool) from 74 % to 97 %.

The code could be complicated at places, because the different
branches in the update_* functions depend on other metadata classes
as well.
Still, I hope we can find a way and simplify the code.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
Ready for review after #1408 has been merged.

sechkova commented on tests/test_trusted_metadata_set.py:
```diff
@@ -118,9 +141,228 @@ def test_update_with_invalid_json(self):
+        invalid_type_data = json.loads(self.metadata["root"])
+        invalid_type_data["signed"]["_type"] = "foo"
+        invalid_type_data = json.dumps(invalid_type_data)
+        # RepositoryError is throwned during new_root deserialization.
```
throwned -> thrown

sechkova commented on tests/test_trusted_metadata_set.py:
```diff
@@ -118,9 +141,228 @@ def test_update_with_invalid_json(self):
+        trusted_set.root_update_finished()
+        with self.assertRaises(RuntimeError):
+            trusted_set.update_root(self.metadata["root"])
+        trusted_set._root_update_finished = False
```
Last line seems redundant.

sechkova commented on tests/test_trusted_metadata_set.py:
```diff
@@ -26,6 +33,19 @@ def setUpClass(cls):
+        cls.timestamp = Metadata.from_bytes(cls.metadata["timestamp"])
+        cls.snapshot = Metadata.from_bytes(cls.metadata["snapshot"])
+        cls.targets = Metadata.from_bytes(cls.metadata["targets"])
+        cls.delegated_role = Metadata.from_bytes(cls.metadata["role1"])
```
I don't see these class variables used anywhere.

MVrachev replied:
They were used, but only once and you are right that if that's the case, then it's better to create those objects close to where they will be used.
Removed those lines.

sechkova commented on tests/test_trusted_metadata_set.py:
```diff
@@ -118,9 +141,228 @@ def test_update_with_invalid_json(self):
+        invalid_type_data["signed"]["_type"] = "foo"
+        invalid_type_data = json.dumps(invalid_type_data)
+        # RepositoryError is throwned during new_root deserialization.
+        # It's not throwned when checking new_root.signed.type != "root"
```
If the goal is to test the case described in the comment too, there should be another test setting
`invalid_root_type_data["signed"]["_type"] = "snapshot"`

MVrachev replied:
Great catch!
Modified my test to use that one instead of "foo".

sechkova commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +137,232 @@ def test_update_with_invalid_json(self):
+        trusted_set.update_timestamp(self.metadata["timestamp"])
+        # cannot update snapshot after targets update completes or targets != None
+        targets_obj = Metadata.from_bytes(self.metadata["targets"])
+        trusted_set._trusted_set["targets"] = targets_obj
```
```suggestion
        trusted_set.targets = targets_obj
```

MVrachev replied:
At the time of the assignment, targets is `None` and that's why I assign it that way because otherwise, I get `AttributeError: can't set attribute`.

sechkova commented on tests/test_trusted_metadata_set.py:
```diff
@@ -118,9 +141,228 @@ def test_update_with_invalid_json(self):
+        data: bytes
+    ) -> Dict[str, str]:
+        modified_hashes = {}
+        # Calculate correct hashes to pass the first check
```
What is "the first check"?

sechkova commented on tests/test_trusted_metadata_set.py:
```diff
@@ -118,9 +141,228 @@ def test_update_with_invalid_json(self):
+            trusted_set.update_snapshot(self.metadata["snapshot"])
+        del trusted_set._trusted_set["targets"]
+
+        #  Deserialization error - failed to load new_snapshot.
```
Maybe a better comment is needed, I'm not sure what is tested here.

sechkova commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +137,232 @@ def test_update_with_invalid_json(self):
+        self._sign_modified_obj("snapshot", snapshot_obj)
+        timestamp_meta = trusted_set.timestamp.signed.meta["snapshot.json"]
+        true_hashes = timestamp_meta.hashes or {}
+        modified_snapshot_data = json.dumps(snapshot_obj.to_dict()).encode()
+        modified_hashes = self._calculate_modified_hashes(
+            true_hashes, json.dumps(snapshot_obj.to_dict()).encode()
+        )
```
```suggestion
        modified_snapshot_data = json.dumps(snapshot_obj.to_dict()).encode()
        modified_hashes = self._calculate_modified_hashes(
            true_hashes, modified_snapshot_data
        )
```

sechkova reviewed (commented):
LGTM with small comments below.

MVrachev commented:
Rebased on top of develop branch to get the latest changes related to hashes and length validation.

Also, I added two new commits:
- the first one `TrustedMetadataSet testing: start using setUp` uses `setUp` to create a trusted metadata set instance per test
- the second one splits the big test functions into smaller ones separated by a test case. This required creating a couple of extra small helpful functions.

Can you have another look @sechkova and tell me do you think it's better that way using many smaller functions?

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
             update_func(metadata)
 
 
+    def test_verify_with_threshold_wrong_delegator(self):
```
let's not add tests for verify_with_threshold(): it's a placeholder until #1436 merges

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+        metadata_obj: Metadata
+    ) -> Dict[str, Any]:
+        key_dict = self.keystore[role]
+        sslib_signer = SSlibSigner(key_dict)
```
self.keystore could be self.signers (`Dict[str, Signer]`) instead so you didn't need to create a Signer every time

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+            verify_with_threshold(targets, "foo", delegated_role)
+
+    def test_update_root_invalid_type(self):
+        # new_root data with invalid snapshot type
+        invalid_type_data = json.loads(self.metadata["root"])
+        invalid_type_data["signed"]["_type"] = "snapshot"
+        invalid_type_data["signed"]["meta"] = {"file1.txt": {"version": 1}}
+        invalid_type_data = json.dumps(invalid_type_data).encode()
+        # RepositoryError is thrown during new_root deserialization.
+        # It's not thrown when checking new_root.signed.type != "root"
+        with self.assertRaises(exceptions.RepositoryError):
+            self.trusted_set.update_root(invalid_type_data)
```
this function can be replaced by 
```
with self.assertRaises(exceptions.RepositoryError):
    self.trusted_set.update_root(self.metadata["snapshot"])
```
correct?

This same test seems to get done in `test_update_with_invalid_json()` already but only for the other top-level metadata, not root... these probably all belong together: would it make sense to move the other "invalid type" tests here as well?

Alternatively, maybe the tests in `test_update_with_invalid_json()` don't actually test what they claim to (because of mismatching metafiles and such...)? if you know that this is the case then feel free to remove the useless ones

MVrachev replied:
I added this test because it was not tested by `test_update_with_invalid_json()`.
At least that's what coverage missing lines tells me.
Will investigate what's the reason.

MVrachev replied:
The reason was that we have forgotten to add this test there.
Will add it there with the rest of the tests testing for invalid root.

jku replied:
what was the resolution here:  is this function needed or not?

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+        with self.assertRaises(exceptions.RepositoryError):
+            self.trusted_set.update_root(invalid_type_data)
+
+    def test_update_root_new_root_cannot_be_verified_with_threshold(self):
+        # new_root data with threshold which cannot be verified.
+        modified_threshold_data = copy.deepcopy(
+            json.loads(self.metadata["root"])
+        )
+        modified_threshold_data["signed"]["roles"]["root"]["threshold"] = 2
+        modified_threshold_data = json.dumps(modified_threshold_data).encode()
+        with self.assertRaises(exceptions.UnsignedMetadataError):
+            self.trusted_set.update_root(modified_threshold_data)
```
the way the test is written leads me to think the threshold value is what's important -- when reality it's just that the signature is no longer valid. Maybe modify signed.version instead -- that's shorter as well?

Also, almost all other test construct a Metadata when they need to do modifications, maybe would be good to be consistent?

jku replied:
what was the resolution here?

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+        self.trusted_set.root_update_finished()
+        with self.assertRaises(RuntimeError):
+            self.trusted_set.update_root(self.metadata["root"])
+        self.trusted_set._root_update_finished = False
```
`test_out_of_order_ops()` checks for these types of errors -- if it does not check for this already, maybe you could add this in the right place there?

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+        )
+        self.trusted_set.root.signed.expires = expired_datetime
+        with self.assertRaises(exceptions.ExpiredMetadataError):
+            self.trusted_set.root_update_finished()
```
you don't need the parsing, something like this works:
`    self.trusted_set.root.signed.expires = datetime(1970, 1, 1)`

The test isn't wrong... but I think it would be better if it actually did load the expired root metadata (to show that the loading works for expired metadata), and then fails on root_update_finished(). The same comment applies to a lot of tests... so we can discuss that in general, no need to do that for this test specifically

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+        timestamp = Metadata.from_bytes(self.metadata["timestamp"])
+        timestamp.signed.expires = datetime.strptime(
+            "1970-01-01T00:00:00Z", "%Y-%m-%dT%H:%M:%SZ"
+        )
```
no need for parsing here either

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+            "1970-01-01T00:00:00Z", "%Y-%m-%dT%H:%M:%SZ"
+        )
+        self._sign_modified_obj("timestamp", timestamp)
+        new_timestamp_byte_data = json.dumps(timestamp.to_dict()).encode()
```
I think this shows we should have a `Timestamp.to_bytes()` to complete the set of de/serialize functions (not in this PR, just commenting)

MVrachev replied:
I think we don't have `to_bytes()` for the rest of the classes as well?
Maybe we can add `Metadata.to_bytes()`?

jku replied:
oh right, that's what I meant (I tried to say `timestamp.to_bytes()` in the example meaning `Metadata.to_bytes()`)

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+        invalid_type_data = json.loads(self.metadata["snapshot"])
+        invalid_type_data["signed"]["_type"] = "targets"
+        invalid_type_data["signed"]["targets"] = {}
+        invalid_type_data = json.dumps(invalid_type_data).encode()
```
is there some advantage to this over just using `self.metadata["targets"]` as the data?

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+        with self.assertRaises(exceptions.RepositoryError):
+            self.trusted_set.update_snapshot(invalid_type_data)
+
+    def test_update_snapshot_after_targets_updated(self):
+        self._setup_update_snapshot_or_timestamp_test()
+        # cannot update snapshot after targets update completes or targets != None
+        targets_obj = Metadata.from_bytes(self.metadata["targets"])
+        self.trusted_set._trusted_set["targets"] = targets_obj
+        with self.assertRaises(RuntimeError):
+            self.trusted_set.update_snapshot(self.metadata["snapshot"])
```
I don't totally understand this:
* Do you need to modify trusted_set internals? That should be absolute last option if nothing else works
* shouldn't this logically fit in `test_out_of_order_ops()` (I thought it would test it already but maybe not)?

MVrachev replied:
I moved this test inside `test_out_of_order_ops` and removed this test function.
It was not added initially in `test_out_of_order_ops`.

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+            self.trusted_set.update_timestamp(new_timestamp_byte_data)
+
+
+    def _calculate_modified_hashes(
+        self, true_hashes,
+        data: bytes
+    ) -> Dict[str, str]:
+        modified_hashes = {}
+        # Calculate hashes on modified data in order to pass hashes verification.
+        for algo in true_hashes.keys():
+            digest_object = sslib_hash.digest(algo)
+            digest_object.update(data)
+            observed_hash = digest_object.hexdigest()
+            modified_hashes[algo] = observed_hash
+        return modified_hashes
```
This method isn't needed? MetaFile does not require hashes so `None` will work fine

MVrachev replied:
Yes, I used what you suggested - assigning hashes and length to None and removed this method.

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+        with self.assertRaises(RuntimeError):
+            self.trusted_set.update_snapshot(self.metadata["snapshot"])
+
+    def test_update_snapshot_with_invalid_json(self):
+        self._setup_update_snapshot_or_timestamp_test()
+        #  Deserialization error - failed to decode the new_snapshot JSON.
+        timestamp_meta = self.trusted_set.timestamp.signed.meta["snapshot.json"]
+        true_hashes = timestamp_meta.hashes or {}
+
+        modified_hashes = self._calculate_modified_hashes(
+            true_hashes, b'{""sig": }'
+        )
+        self.trusted_set.timestamp.signed.meta["snapshot.json"].hashes = modified_hashes
+        with self.assertRaises(exceptions.RepositoryError):
+            self.trusted_set.update_snapshot(b'{""sig": }')
+        self.trusted_set.timestamp.signed.meta["snapshot.json"].hashes = true_hashes
+
```
can you explain better what this tests.

If it's testing the metafile verification in TrustedMetadataSet, I think it's easier (much easier for the reader) if you just modify metafile.version

If it's testing some specific failure within MetaFile verify, then the test probably doesn't belong in TrustedMetadataSet tests.

MVrachev replied:
I moved this test into `test_out_of_order_ops`.
It's different and easier to understand now.

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+        self._setup_update_snapshot_or_timestamp_test()
+        self.trusted_set.update_snapshot(self.metadata["snapshot"])
+
+    def test_update_snapshot_after_successful_update_new_snapshot_no_meta(self):
+        self._setup_update_snapshot_after_successful_update_test()
+        # Test removing a meta_file in new_snapshot compared to the old snapshot
+        snapshot_obj = Metadata.from_bytes(self.metadata["snapshot"])
+        snapshot_obj.signed.meta = {}
+        # prepare timestamp.meta["snapshot"].hashes
+        self._sign_modified_obj("snapshot", snapshot_obj)
+        timestamp_meta = self.trusted_set.timestamp.signed.meta["snapshot.json"]
+        true_hashes = timestamp_meta.hashes or {}
+        modified_snapshot_data = json.dumps(snapshot_obj.to_dict()).encode()
+        modified_hashes = self._calculate_modified_hashes(
+            true_hashes, modified_snapshot_data
+        )
+        self.trusted_set.timestamp.signed.meta["snapshot.json"].hashes = modified_hashes
```
this is quite difficult to understand. Any chance of simplifying (e.g. not using hashes)?

MVrachev replied:
Yea, I simplified it without calculating hashes.

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+    def test_update_snapshot_after_succesfull_update_new_snapshot_meta_version_different(self):
+        self._setup_update_snapshot_after_successful_update_test()
+        # snapshot.meta["project1"].version != new_snapshot.meta["project1"].version
+        for meta_file_path in self.trusted_set.snapshot.signed.meta.keys():
+            self.trusted_set.snapshot.signed.meta[meta_file_path].version = 2
```
```suggestion
for metafile in self.trusted_set.snapshot.signed.meta.values():
    metafile.version += 1
```
or something like it?

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+        self._setup_update_snapshot_after_successful_update_test()
+        # new_snapshot has expired
+        snapshot_obj = Metadata.from_bytes(self.metadata["snapshot"])
+        snapshot_obj.signed.expires = datetime.strptime(
+            "1970-01-01T00:00:00Z", "%Y-%m-%dT%H:%M:%SZ"
+        )
```
unneeded parsing

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+            self.trusted_set.update_snapshot(modified_snapshot_data)
+
+
+    def setup_update_targets_test(self):
```
* other setup function names start with an underscore
* Maybe the setup functions could also all be close to each other so it's easier to see the differences?

MVrachev replied:
Yes, I will group them together just after `setUp()`.

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+        with self.assertRaises(exceptions.RepositoryError):
+            self.trusted_set.update_targets(self.metadata["targets"])
+
+    def test_update_targets_hash_different_than_snapshot_meta_hash(self):
+        self.setup_update_targets_test()
+        # observed_hash != stored hash in snapshot meta for targets
+        true_hashes = {}
+        for target_path, meta_file in self.trusted_set.snapshot.signed.meta.items():
+            true_hashes[target_path] = meta_file.hashes
+            self.trusted_set.snapshot.signed.meta[target_path].hashes = {"sha256": "b"}
+        with self.assertRaises(exceptions.RepositoryError):
+            self.trusted_set.update_targets(self.metadata["targets"])
```
true_hashes is unused

Also I think mentioning hashes in the test name is maybe not what we want: TrustedMetadataSet is not interested in whether the issue is hashes, version or length: we're testing that TrustedMetadataSet functions correctly when MetaFile verification fails (for any reason)

MVrachev replied:
Yes, but I want to be specific about what's wrong with this particular test because we have around 15 of them.

jku replied:
> true_hashes is unused

MVrachev replied:
Thanks, I missed that.

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+        with self.assertRaises(exceptions.RepositoryError):
+            self.trusted_set.update_targets(self.metadata["targets"])
+
+    def test_update_targets_version_different_snapshot_meta_version(self):
```
I think this is the same test as above (see the comment there).

I _think_ the exception is now the same as well, after some of of Teodoras PRs...

MVrachev replied:
From what I understand, they are different. There is a separate exception raised for bad versions.

jku replied:
you are right: I mistook these for testing length and hashes, but it's version and hashes (and version does get tested much later in the process)

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
+        self.setup_update_targets_test()
+        # new_delegated_target has expired
+        targets_obj = Metadata.from_bytes(self.metadata["targets"])
+        targets_obj.signed.expires = datetime.strptime(
+            "1970-01-01T00:00:00Z", "%Y-%m-%dT%H:%M:%SZ"
+        )
```
unneeded parsing

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -110,13 +128,269 @@ def test_update_with_invalid_json(self):
             # metadata is of wrong type
             with self.assertRaises(exceptions.RepositoryError):
-                update_func(data["root"])
+                update_func(self.metadata["root"])
```
Based on the tests you wrote I guess this does not actually test what I thought it would -- for snapshot and targets this is going to fail in metafile verification already, not because the type is wrong ... feel free to remove this if you have the root/timestamp cases covered elsewhere

MVrachev replied:
This won't work only for snapshot where there are hashes and length verification.
I will add commit to handing this case.

jku reviewed (changes requested):
Sooo many comments, sorry about that... but it is a big patch and again the first of it's kind.

I like the smaller function, much easier to read.

many tests finish by modifying some trusted_set metadata internals: that should never be needed, or am I missing something?

There's a lot of hash calculating that I think isn't needed: MetaFile hashes are optional, you can just remove the hashes if they are there.

Many tests follow this pattern:
* update TrustedMetadataSet using valid metadata
* modify TrustedMetadataSet internal metadata
* do the test

In some cases that's what you have to do... but in a lot of cases I think it might be beneficial if we could avoid modifying internal state and instead could do:
* modify metadata before feeding it to TrustedMetadataSet
* Update TrustedMetadataSet with modified metadata (triggering the test)

It's possible that this change is not worth the trouble or should be in another PR... but let's consider at least

MVrachev commented:
I addressed your comments @jku.

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -1,122 +1,330 @@
+from tuf.ngclient._internal.trusted_metadata_set import(
+    TrustedMetadataSet
+)
+from securesystemslib import hash as sslib_hash
```
unused?

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -1,122 +1,330 @@
+    def _setup_update_targets_test(self):
+        self.trusted_set.root_update_finished()
+        self.trusted_set.update_timestamp(self.metadata["timestamp"])
+        self.trusted_set.update_snapshot(self.metadata["snapshot"])
```
now that I see them next to each other, these two look like the same function?

It might be more useful if the functions were named for what they do, not what is expected to happen after the function.

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -1,122 +1,330 @@
+            self.metadata["role1"], "role1", "targets"
+        )
+
+        self.trusted_set.update_targets(self.metadata["targets"])
```
wait, what is this line trying to test for? 

How does it succeed, is this a TrustedMetadataSet bug?

jku replied:
I think this is just an unneeded call...

The call succeeds (even though delegated target role1 has already been loaded) against what's said in the TrustedMetadataSet docs _"Metadata is not updatable if any metadata depending on it has been loaded"_. We ended up with a version that does not explicitly check that I think because the meta info in snapshot should ensure that the targets metadata is consistent even if a delegating metadata gets "reloaded". That said, there is now a possibility of a repository doing weird things like publishing a new "targets.json" with same version but modified delegations... maybe it's not something we need to care about, but at least the TrustedMetadataSet docs should be updated  (not in this PR)

jku replied:
actually the quote from the docs is for top-level metadata so it is technically correct... Might make sense to mention that delegating targets consistency comes from snapshots meta information.

MVrachev replied:
> That said, there is now a possibility of a repository doing weird things like publishing a new "targets.json" with same version but modified delegations... maybe it's not something we need to care about, but at least the TrustedMetadataSet docs should be updated (not in this PR)

I guess we don't care about that because if the file is signed from the keys trusted by root.json then everything is fine right? 

We haven't documented which roles are used for verification when calling all other `update_*` functions.
Why do we want to add this specifically here? 

PS: I will remove the redundant two updates of `update_targets` and `update_delegated_targets`.

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -1,122 +1,330 @@
+
+    def test_update_snapshot_cannot_verify_snapshot_with_threshold(self):
+        self._setup_update_snapshot_or_timestamp_test()
+        # remove signature for snapshot from root data
+        self.trusted_set.root.signed.roles["snapshot"].keyids = []
+        # self.trusted_set.snapshot.signatures = {}
```
what is going on here?

MVrachev replied:
Not sure how this line slipped. It should have happened during a couple of releases I had to do.
Thanks for catching this!

jku reviewed (commented):
I'm seeing several smaller issues in the code that I've commented on and discussion was marked resolved. I'm not sure how to interpret those. I guess I'll unresolve. Please try to comment in the discussions if you don't think fixes are needed.

Do you have an opinion on the pattern I mentioned (modifying TrustedMetadataSet internals during the test): are there solutions to avoid that, what would they look like, do we want to try them or leave for later?

MVrachev commented:
> I'm seeing several smaller issues in the code that I've commented on and discussion was marked resolved. I'm not sure how to interpret those. I guess I'll unresolve. Please try to comment in the discussions if you don't think fixes are needed.

I am sorry that I closed discussions that weren't addressed.
I rebased a couple of times locally and I was distracted.

> Do you have an opinion on the pattern I mentioned (modifying TrustedMetadataSet internals during the test): are there solutions to avoid that, what would they look like, do we want to try them or leave for later?

I have to experiment with this and will do that on Monday.
My only concern is that it could be more complicated as most of the time a test depends on the state of other metadata classes inside the set. 

So, for example I want to check if `new_snapshot.version != trusted timestamp.meta["snapshot"].version` will fail.
I will have to do the following operations:

To do it the way you propose I will have to:
1.  instantiate a timestamp object from bytes
2. change the timestamp_obj.meta["snapshot"].version
3. convert timestamp_obj to bytes again
4. call `update_timestamp`
5. perform the test

Compared to now where I do:
1. call `update_timestamp`
2. change the internal `timestamp.signed.meta["snapshot.json"].version`
3. perform the test.

PS: I have submitted https://github.com/theupdateframework/tuf/pull/1490 which I hope will simplify the code a little.

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -1,122 +1,319 @@
+            self.trusted_set.update_timestamp(new_timestamp_byte_data)
+
+
+    def test_update_snapshot_after_targets_updated(self):
+        self._root_update_finished_and_update_timestamp()
+        # cannot update snapshot after targets update completes or targets != None
+        targets_obj = Metadata.from_bytes(self.metadata["targets"])
+        self.trusted_set._trusted_set["targets"] = targets_obj
+        with self.assertRaises(RuntimeError):
+            self.trusted_set.update_snapshot(self.metadata["snapshot"])
```
I think you've added this test in test_out_of_order_ops() already?

(without any need to modify internal state, I might add)

MVrachev replied:
Ohh.. I am not sure how I forgot to remove it.
Thanks!

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -1,122 +1,319 @@
+        self.trusted_set.root.signed.roles["snapshot"].keyids = []
+        with self.assertRaises(exceptions.UnsignedMetadataError):
+            self.trusted_set.update_snapshot(self.metadata["snapshot"])
+        self.trusted_set.root.signed.roles["snapshot"].threshold = 1
```
The last assignment seems unrelated.

nitpick: comment is still weird (talks about signature but removes keys from snapshot role).

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -1,122 +1,319 @@
+    ) -> Dict[str, Any]:
+        sslib_signer = self.keystore[role]
+        signature = metadata_obj.sign(sslib_signer)
+        return signature.to_dict()
```
unused return value.

is this function valuable? it seems replaceable by  e.g. `root_obj.sign(self.keystore["root"])` (example from a few lines below)

MVrachev replied:
Yes, you are right.
Now when we have assigned SSLib instances to `self.keystore` it makes sense to remove this helper function.

jku reviewed (commented):
> To do it the way you propose I will have to:
> 
> 1. instantiate a timestamp object from bytes
> 2. change the timestamp_obj.meta["snapshot"].version
> 3. convert timestamp_obj to bytes again
> 4. call `update_timestamp`
> 5. perform the test

You make it sound like this is  a lot more complex but I'm not sure it is 

For example, the current test: 

```python
    def test_update_snapshot_cannot_verify_snapshot_with_threshold(self):
        self._root_update_finished_and_update_timestamp()
        # remove signature for snapshot from root data
        self.trusted_set.root.signed.roles["snapshot"].keyids = []
        with self.assertRaises(exceptions.UnsignedMetadataError):
            self.trusted_set.update_snapshot(self.metadata["snapshot"])
        self.trusted_set.root.signed.roles["snapshot"].threshold = 1
```

could be something like 

```python
    def test_update_snapshot_cannot_verify_snapshot_with_threshold(self):
        self._root_update_finished_and_update_timestamp()

        md = Metadata.from_bytes(self.metadata["snapshot"])
        md.signatures.clear()

        with self.assertRaises(exceptions.UnsignedMetadataError):
            self.trusted_set.update_snapshot(JSONSerializer().serialize(md))
```
That does not look complex in comparison.

Sure, in some cases you need to sign the resulting metadata as well: I'm sure there could be a helper function that makes the whole edit process reasonable for individual tests.

I'm **really** not saying we need to do this now though -- this PR is big enough, feel free to leave it as is to move things along. 

I am saying that as general rule, tests are better when they do not modify internal state: we will likely keep refactoring the implementation and then slowly we will have less trust on the tests (as they rely on the internals that have now changed)

Commit:
TrustedMetadataSet testing: use Metadata.to_bytes

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
ngcl. tests: remove modification on internal state

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
@jku I decide to experiment with updating the tests so that we don't change the internal state of the trusted metadata set.

I was wondering do I want to call `self._root_updated_and_update_timestamp(self._modify_timestamp_meta(version=3))`
directly or instead do it like I did in the tests
```
modified_timestamp = self._modify_timestamp_meta(version=3)
self._root_updated_and_update_timestamp(modified_timestamp)
```

What do you think of the way I implemented it? 
Does the new helper function look understandable?

Commit:
Embed modification function helpers

Instead of using general abstract modification functions embed smaller
modification functions inside each test where it's needed and
create modify_metadata function that does all of the common stuff like:
- instantiating a metadata object
- calling the modification function
- signing the modified object
- serializing back to bytes.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
As suggested by @jku, in my last commit I removed the `modify_timestamp` and `modify_snapshot` functions and instead defined `modify_metadata(rolename: str, modification_func: Callable[["Signed"], None])` function which does the common thinks
like:
- instantiating a metadata object
- calling the modification function
- signing the modified object
- serializing back to bytes

jku commented:
I think that's a pretty good solution with the constraints you have... It's not immediately obvious how modify_metadata() works but reader can figure it out and it makes test code very readable.

The one suggestion I still have is about hashes and length in timestamp: I understand you don't want to modify the test json file but you could just set hashes and length to None in setUpClass() once. That would simplify code in 9 different tests.

I'll approve this as is, you decide.

jku reviewed (approved):
Approved, see comment above.

Commit:
Annotations and use extensively modify_metadata

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
@jku I applied your suggestion in my latest commit and additionally noticed there are more places where I can simplify the code or add annotations.
